README.md
fridumpX (v1.1) is an open source memory dumping tool and using regex patterns (verify regex.yaml) to grep sensitive information based on these regexs, primarily aimed to **penetration testers and developers** to verify memory information that contains some sensite information at memory, all those information based on regex patterns ( you can find at regex.yaml file). 

FridumpX uses Frida framework to dump accessible memory addresses from any platform supported. It can be used from a Windows, Linux or Mac OS X systems to dump the memory of an iOS, Android or Windows application.

---

* **FridumpX** was tested with:

      Frida 16.0.8;
      iOS jailbreaked devices with: 
            - https://checkra.in;
            - https://github.com/palera1n;
            - Note, for iphone X > 14.0 use Xina jailbreak.

* Regex used

      - http and https urls
      - amazon tokens
      - gcp tokens
      - aws 
      - paypal
      - and every regex u want just insert at regex.yaml file

---


### Pocs

![fridumpx02.gif](fridumpx02.gif)
![fridumpx01.gif](fridumpx01.gif)
![fridumpX.png](fridumpX.png)
![fridumpX.png](fridumpX1.png)
![fridumpX.png](fridumpX2.png)


Usage
---
```
usage: fridumpX [-h] [-re] [-reo dir] [-o dir] [-U] [-v] [-r] [-s] [--max-size bytes] [-D id] [-H host] [-dnl] [-cdnl CUSTOMDENYLIST] process

Author: @unp4ck
███████ ██████  ██ ██████  ██    ██ ███    ███ ██████  ██   ██
██      ██   ██ ██ ██   ██ ██    ██ ████  ████ ██   ██  ██ ██
█████   ██████  ██ ██   ██ ██    ██ ██ ████ ██ ██████    ███
██      ██   ██ ██ ██   ██ ██    ██ ██  ██  ██ ██       ██ ██
██      ██   ██ ██ ██████   ██████  ██      ██ ██      ██   ██

Powered by: @fridadotre
Version: v1.1

positional arguments:
  process               the process that you will be injecting to

options:
  -h, --help            show this help message and exit
  -re, --regex          use regex to grep juicy info, use the -reo option to save the findings
  -reo dir, --regexout dir
                        dir name after regex findings (def: fridumpx_out
  -o dir, --out dir     provide full output directory path. (def: 'dump')
  -U, --usb             device connected over usb
  -v, --verbose         verbose
  -r, --read-only       dump read-only parts of memory. More data, more errors
  -s, --strings         run strings on all dump files. Saved in output dir.
  --max-size bytes      maximum size of dump file in bytes (def: 20971520)
  -D id, --device id    connect to device with the given id
  -H host, --host host  remote device connected over tcp
  -dnl, --denylist      common patterns to deny while using the regex feature
  -cdnl CUSTOMDENYLIST, --customdenylist CUSTOMDENYLIST
                        custom deny list with comma separated values ex: foo,bar,john,doe
```

#### How to

To find the name of a local process, you can use:

      frida-ps
For a process that is running on a USB connected device, you can use:

      frida-ps -U

Examples:


* Mobile app target, dump the memory and run strings on all dump files (xxxxx.data):

```bash
python3 fridumpX.py -U com.app.name -s

# verify at dump folder, you can cat all .data files
$ cat *.data | grep -i "<your_pattern>" 
```

* Any target, Dump the memory of a local application and save it to the specified directory

```
python3 fridumpX -r -o [full_path]
``` 

* Any target, Dump the memory and running regex of a local application files and save it to the specified directory

```
python3 fridumpX -U app.name -re -reo app_target_folder
``` 

* Any target, Dump the memory and running regex of a local application files and save it to the specified directory with DENYLIST

```
python3 fridumpX -U app.name -re -reo app_target_folder -dnl
```

* Deny List and Custom Deny List

```
python3 fridumpX.py -U com.apple.mobiletimer -s -re -reo timerApple -dnl -cdnl apple,microsoft,google
```

      
#### Installation
---
To install fridumpX you just need to clone it from git and run it:

      git clone https://github.com/unp4ck/fridumpX.git

      $ python3 -m pip install -r requirements.txt --user
            
      python3 fridumpX.py -h
            
Pre-requisites
---
To use fridumpX you need to have frida installed on your python environment and frida-server on the device you are trying to dump the memory from.
The easiest way to install frida on your python is using pip:

```

python3 -m install frida frida-tools

```

## FAQ

 This tool are inspired/forked by fridump.

* https://github.com/frida/frida            thanks to @frida
* https://github.com/Nightbringer21/fridump thanks to @Nßightbringer
    
* More information on how to install Frida can be found [here](http://www.frida.re/docs/installation/)

* For iOS, installation instructions can be found [here](http://www.frida.re/docs/ios/).

* For Android, installation instructions can be found [here](http://www.frida.re/docs/android/).

Any suggestions and comments are welcome! @unp4ck

### Contribute

Before open a issue, verify if:

* 1 - Your app have protections like: (anti(root|jailbreak|tamper|sslpinning);
* 2 - Verify if your app targer working correct, try to upgrade your frida to 16.x or 15.x;
* 3 - Verify your frida-tools and frida-version, we not tested versions bellow 14.x;
* 4 - Before SEND a issue verify the steps above and, if possible send your testing app, we cannot simulate your problem without steps to reproduce


### TODO

* 1 - [main task] - Verify why some apps are slow, to running strings grepper!
* 2 - Bypass JB/Root with frida scripts
* 3 - Code Refactory
* 4 - https://blog.csdn.net/freeking101/article/details/107438271



CHANGELOG.md
### v1.0

* Fixing script issues while parsing memory
* Logic Erros


### v1.1

* Add Spawn delay
* Fixing some spawn problems with frida python

fridawizz.py
import textwrap
import frida
import os
import sys
import frida.core
import core.dumper as dumper
import argparse
import logging
import time
from utils import misc, Utils




# Main Menu
def MENU():
    parser = argparse.ArgumentParser(
        prog='fridawizz',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description=textwrap.dedent(""))

    parser.add_argument('process',
                        help='the process that you will be injecting to')
    parser.add_argument('-o', '--out', type=str, metavar="dir",
                        help='provide full output directory path. (def: \'dump\')')
    parser.add_argument('-U', '--usb', action='store_true',
                        help='device connected over usb')
    parser.add_argument('-v', '--verbose', action='store_true',
                        help='verbose')
    parser.add_argument('-r', '--read-only', action='store_true',
                        help="dump read-only parts of memory. More data, more errors")
    parser.add_argument('-s', '--strings', action='store_true',
                        help='run strings on all dump files. Saved in output dir.')
    parser.add_argument('--max-size', type=int, metavar="bytes",
                        help='maximum size of dump file in bytes (def: 20971520)')
    parser.add_argument('-D', '--device', type=str, metavar='id',
                        help='connect to device with the given id')
    parser.add_argument('-H', '--host', type=str, metavar="host", 
                        help='remote device connected over tcp') 
    args = parser.parse_args()
    return args


arguments = MENU()
print(misc.logo)

# Define Configurations
APP_NAME = arguments.process
DIRECTORY = ""
USB = arguments.usb
DEBUG_LEVEL = logging.INFO
STRINGS = arguments.strings
MAX_SIZE = 20971520 # max integer value python
PERMS = 'rw-'
DEVICE = arguments.device
HOST = arguments.host # TODO

if arguments.read_only:
    PERMS = 'r--'

if arguments.verbose:
    DEBUG_LEVEL = logging.DEBUG
logging.basicConfig(format='%(levelname)s:%(message)s', level=DEBUG_LEVEL)


# Start a new Session


session = None

try:
    if USB:
        # if len(HOST) > 0:
        #     device_manager = frida.get_device_manager()
        #     device = device_manager.add_remote_device(HOST)
        #     pid = device.spawn([APP_NAME])
        #     device.resume(pid)
        #     session = frida.get_usb_device().attach(pid)
        device =  frida.get_usb_device()
        pid = device.spawn([APP_NAME])
        device.resume(pid)
        time.sleep(3) # LOL some apps ;p
        session = frida.get_usb_device().attach(pid)
    elif DEVICE:
        session = frida.get_device(DEVICE).attach(APP_NAME)
    else:
        session = frida.attach(APP_NAME)
except Exception as e:
    print("Can't connect to App. Have you connected the device?")
    logging.debug(str(e))
    sys.exit()


# Selecting Output directory
if arguments.out is not None:
    DIRECTORY = arguments.out
    if os.path.isdir(DIRECTORY):
        print("Output directory is set to: " + DIRECTORY)
    else:
        print("The selected output directory does not exist!")
        sys.exit(1)

else:
    print("Current Directory: " + str(os.getcwd()))
    DIRECTORY = os.path.join(os.getcwd(), "dump")
    print("Output directory is set to: " + DIRECTORY)
    if not os.path.exists(DIRECTORY):
        print("Creating directory...")
        os.makedirs(DIRECTORY)

mem_access_viol = ""

print("Starting Memory dump...")

script = session.create_script(
    """'use strict';

    rpc.exports = {
      enumerateRanges: function (prot) {
        return Process.enumerateRangesSync(prot);
      },
      readMemory: function (address, size) {
        return Memory.readByteArray(ptr(address), size);
      }
    };

    """)
script.on("message", Utils.on_message)
script.load()

agent = script.exports
ranges = agent.enumerate_ranges(PERMS)

if arguments.max_size is not None:
    MAX_SIZE = arguments.max_size

i = 0
l = len(ranges)

# Performing the memory dump
for range in ranges:
    base = range["base"]
    size = range["size"]

    logging.debug("Base Address: " + str(base))
    logging.debug("")
    logging.debug("Size: " + str(size))

    if size > MAX_SIZE:
        logging.debug("Too big, splitting the dump into chunks")
        mem_access_viol = dumper.splitter(
            agent, base, size, MAX_SIZE, mem_access_viol, DIRECTORY)
        continue
    mem_access_viol = dumper.dump_to_file(
        agent, base, size, mem_access_viol, DIRECTORY)
    i += 1
    Utils.printProgress(i, l, prefix='Progress:', suffix='Complete', bar=50)
print("")

# Run Strings if selected

if STRINGS:
    files = os.listdir(DIRECTORY)
    i = 0
    l = len(files)
    print("Running strings on all files:")
    for f1 in files:
        Utils.strings(f1, DIRECTORY)
        i += 1
        Utils.printProgress(i, l, prefix='Progress:', suffix='Complete', bar=50)
    aws = input("[*] Would you like to find juicy information wih regex patterns: (Y/n)")
    if aws.lower()[0] == 'y':
        Utils.inspec(DIRECTORY)
    else:
        print("[*] Thanks for using this tool! <3")
        print("Finished!")
        sys.exit(0)

core/dumper.py
import os
import logging
import sys
import hashlib

# Reading bytes from session and saving it to a file

def dump_to_file(agent,base,size,error,directory):
        try:
                filename = str(base)+'_dump.data'
                dump =  agent.read_memory(base, size)
                f = open(os.path.join(directory,filename), 'wb')
                f.write(dump)
                f.close()
                return error
        except Exception as e:
            logging.debug("[!]"+str(e))
            print("[*] memory access violation!")
            return error

def regex_dump_file(directory, finding):
    try:
        with open(os.path.join(directory,'regex_findings.txt'), 'a') as f:
            write_line = finding + '\n'
            f.writelines(write_line)
    except Exception as e:
        logging.debug(f"[!] {e}")
        print(f"[!] error while saving regex findings {e}")
        sys.exit(1)
    

    ## remove duplicates
    input_file_path = os.path.join(directory,'regex_findings.txt')
    out_file_path = os.path.join(directory,'regex_findings_uniq.txt')

    complete_line_hash = set()
    out_file = open(out_file_path, "w")
    for line in open(input_file_path, "r"):
        hash_value = hashlib.md5(line.rstrip().encode('utf-8')).hexdigest()
        if hash_value not in complete_line_hash:
            out_file.write(line+'\n')
            complete_line_hash.add(hash_value)
    out_file.close()



#Read bytes that are bigger than the max_size value, split them into chunks and save them to a file

def splitter(agent,base,size,max_size,error,directory):
        #times = size//max_size
        times = int(size/max_size) #fix where times is not a number
        diff = size % max_size
        if diff is 0:
            logging.debug("Number of chunks:"+str(times+1))
        else:
            logging.debug("Number of chunks:"+str(times))
        global cur_base
        cur_base = int(base,0)

        for time in range(int(times)):
                logging.debug("Save bytes: "+str(cur_base)+" till "+str(cur_base+max_size))
                dump_to_file(agent, cur_base, max_size, error, directory)
                cur_base = cur_base + max_size

        if diff is not 0:
            logging.debug("Save bytes: "+str(hex(cur_base))+" till "+str(hex(cur_base+diff)))
            dump_to_file(agent, cur_base, diff, error, directory)



utils/Utils.py
import sys
import string
import logging
import os
import re
from io import open
import yaml
from core import dumper
from urllib.parse import urlparse


def read_yaml(file_path):
    with open(file_path, "r") as f:
        return yaml.safe_load(f)

def inspec(directory, regex_dir_name, denyListCustom="", DENY_ON=False):
    # Use os.listdir() to get a list of all the files in the directory
    files = os.listdir(directory)
    # Iterate over the list of files and read each one
    for file in files:
        if os.path.isdir(file):
            pass
        with open(os.path.join(directory, file), 'r', encoding='iso8859-1') as f:
            # print(f"[*] reading file: {file}") uncoment if you want to debug :}
            content = f.read()
            #print(content)
            # open regex file
            regex_file = read_yaml('regex.yaml')
            for key, value in regex_file.items():
                regex_obj = re.compile(value)
                check = regex_obj.findall(content)
                if check:
                    for findings in set(check):
                        if DENY_ON:
                            if key == "HTTP" or key == "HTTPS":
                                if denyURL(findings):
                                    continue
                                else:
                                    print(f"[*] FOUND pattern: [{key}]:     {findings}")
                                    dumper.regex_dump_file(regex_dir_name, findings)
                        if DENY_ON and denyListCustom != "":
                            if denyListCustom != "" or denyListCustom != None:
                                if denyListLogic(findings, denyListCustom):
                                    continue
                                else:
                                    print(f"[*] FOUND pattern: [{key}]:     {findings}")
                                    dumper.regex_dump_file(regex_dir_name, findings)
                            else:
                                print(f"[*] FOUND pattern: [{key}]:     {findings}")
                                dumper.regex_dump_file(regex_dir_name, findings)



def denyURL(url):
    with open('denylist.txt') as denyListTXT:
       lines = [line.rstrip() for line in denyListTXT]
       for line in lines:
            parseURL = urlparse(url)
            if parseURL.hostname == line:
                # print(f"{parseURL.hostname} == {line} || {parseURL.hostname == line}")
                return True

def denyListLogic(stringFound, denyListCustom=""):
    if denyListCustom != "" or denyListCustom != None:
        for s in denyListCustom.split(","):
            if s in stringFound:
                return True


# Progress bar function
def printProgress(times, total, prefix='', suffix='', decimals=2, bar=100):
    filled = int(round(bar * times / float(total)))
    percents = round(100.00 * (times / float(total)), decimals)
    bar = '#' * filled + '-' * (bar - filled)
    sys.stdout.write('%s [%s] %s%s %s\r' %
                     (prefix, bar, percents, '%', suffix)),
    sys.stdout.flush()
    if times == total:
        print("\n")


# A very basic implementations of Strings
def strings(filename, directory, min=4):
    strings_file = os.path.join(directory, "strings.txt")
    path = os.path.join(directory, filename)
    with open(path, encoding='Latin-1') as infile:
        str_list = re.findall("[\x20-\x7E]+\x00", infile.read())
        with open(strings_file, "a") as st:
            for string in str_list:
                if len(string) > min:
                    logging.debug(string)
                    st.write(string + "\n")

# Method to receive messages from Javascript API calls


def on_message(message, data):
   print("[*] [on_message] message:", message, "data:", data)


fridumpX.py
import textwrap
import frida
import os
import sys
import frida.core
import core.dumper as dumper
import argparse
import logging
import time
from utils import misc, Utils


__author__ = "@unp4ck"
__version__ = "v1.1"

# Main Menu
def MENU():
    parser = argparse.ArgumentParser(
        prog='fridumpX',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description=textwrap.dedent(misc.logo(__version__)))

    parser.add_argument('process',
                        help='the process that you will be injecting to')
    parser.add_argument('-re', '--regex', action='store_true', help='use regex to grep juicy info, use the -reo option to save the findings')
    parser.add_argument('-reo', '--regexout', type=str, metavar="dir", help="dir name after regex findings (def: fridumpx_out")
    parser.add_argument('-o', '--out', type=str, metavar="dir",
                        help='provide full output directory path. (def: \'dump\')')
    parser.add_argument('-U', '--usb', action='store_true',
                        help='device connected over usb')
    parser.add_argument('-v', '--verbose', action='store_true',
                        help='verbose')
    parser.add_argument('-r', '--read-only', action='store_true',
                        help="dump read-only parts of memory. More data, more errors")
    parser.add_argument('-s', '--strings', action='store_true',
                        help='run strings on all dump files. Saved in output dir.')
    parser.add_argument('--max-size', type=int, metavar="bytes",
                        help='maximum size of dump file in bytes (def: 20971520)')
    parser.add_argument('-D', '--device', type=str, metavar='id',
                        help='connect to device with the given id')
    parser.add_argument('-H', '--host', type=str, metavar="host", 
                        help='remote device connected over tcp')
    parser.add_argument('-dnl', '--denylist', help="common patterns to deny while using the regex feature", action='store_true')
    parser.add_argument('-cdnl', '--customdenylist', help='custom deny list with comma separated values ex: foo,bar,john,doe', type=str)
    args = parser.parse_args()
    return args


arguments = MENU()
print(misc.logo(__version__))

# Define Configurations
APP_NAME = arguments.process
DIRECTORY = ""
USB = arguments.usb
DEBUG_LEVEL = logging.INFO
STRINGS = arguments.strings
MAX_SIZE = 20971520 # max integer value python
PERMS = 'rw-'
DEVICE = arguments.device
HOST = arguments.host # TODO
REGEX = arguments.regex
REGEX_DIR = arguments.regexout
REGEX_DIR_NAME = ""
DENY_LIST_CUSTOM = arguments.customdenylist
DENY_ON = arguments.denylist

if arguments.read_only:
    PERMS = 'r--'

if arguments.verbose:
    DEBUG_LEVEL = logging.DEBUG
logging.basicConfig(format='%(levelname)s:%(message)s', level=DEBUG_LEVEL)


# Start a new Session


session = None

try:
    if USB:
        # if len(HOST) > 0:
        #     device_manager = frida.get_device_manager()
        #     device = device_manager.add_remote_device(HOST)
        #     pid = device.spawn([APP_NAME])
        #     device.resume(pid)
        #     session = frida.get_usb_device().attach(pid)
        device =  frida.get_usb_device()
        pid = device.spawn([APP_NAME])
        device.resume(pid)
        time.sleep(3) # LOL some apps ;p
        session = frida.get_usb_device().attach(pid)
    elif DEVICE:
        session = frida.get_device(DEVICE).attach(APP_NAME)
    else:
        session = frida.attach(APP_NAME)
except Exception as e:
    print("Can't connect to App. Have you connected the device?")
    logging.debug(str(e))
    sys.exit()


# Selecting Output directory
if arguments.out is not None:
    DIRECTORY = arguments.out
    if os.path.isdir(DIRECTORY):
        print("Output directory is set to: " + DIRECTORY)
    else:
        print("The selected output directory does not exist!")
        sys.exit(1)

else:
    print("Current Directory: " + str(os.getcwd()))
    DIRECTORY = os.path.join(os.getcwd(), "dump")
    print("Output directory is set to: " + DIRECTORY)
    if not os.path.exists(DIRECTORY):
        print("Creating directory...")
        os.makedirs(DIRECTORY)

mem_access_viol = ""

print("Starting Memory dump...")

script = session.create_script(
    """'use strict';

    rpc.exports = {
      enumerateRanges: function (prot) {
        return Process.enumerateRangesSync(prot);
      },
      readMemory: function (address, size) {
        return Memory.readByteArray(ptr(address), size);
      }
    };

    """)
script.on("message", Utils.on_message)
script.load()

agent = script.exports
ranges = agent.enumerate_ranges(PERMS)

if arguments.max_size is not None:
    MAX_SIZE = arguments.max_size

i = 0
l = len(ranges)

# Performing the memory dump
for range in ranges:
    base = range["base"]
    size = range["size"]

    logging.debug("Base Address: " + str(base))
    logging.debug("")
    logging.debug("Size: " + str(size))

    if size > MAX_SIZE:
        logging.debug("Too big, splitting the dump into chunks")
        mem_access_viol = dumper.splitter(
            agent, base, size, MAX_SIZE, mem_access_viol, DIRECTORY)
        continue
    mem_access_viol = dumper.dump_to_file(
        agent, base, size, mem_access_viol, DIRECTORY)
    i += 1
    Utils.printProgress(i, l, prefix='Progress:', suffix='Complete', bar=50)
print("")

# Run Strings if selected

if REGEX and not STRINGS:
    print("[!] you need to insert -s\n python3 fridumpx.py -s com.foo.bar -reo foo.bar -re")
    print("[*] exiting..")
    

if STRINGS:
    files = os.listdir(DIRECTORY)
    i = 0
    l = len(files)
    print("[*] Running strings on all files!")
    for f1 in files:
        Utils.strings(f1, DIRECTORY)
        i += 1
        Utils.printProgress(i, l, prefix='Progress:', suffix='Complete', bar=50)
    if REGEX:
        if REGEX_DIR is not None:
            REGEX_DIR_NAME = REGEX_DIR
            if not os.path.exists(REGEX_DIR_NAME):
                os.makedirs(REGEX_DIR_NAME)
        if os.path.isdir(REGEX_DIR_NAME):
            print("[*] Output regexs finding directory is set to: " + REGEX_DIR_NAME)
            print(f"[*] dumping contents of {DIRECTORY}")
            Utils.inspec(DIRECTORY, REGEX_DIR_NAME, DENY_LIST_CUSTOM, DENY_ON)
        else:
            print("[!] The selected output directory does not exist!")
            sys.exit(1)
    else:
        print("[*] Thanks for using this tool! <3")
        print("Finished!")
        sys.exit(0)